File: src/AetherMesh.Storage/FileSystemKeyValueStore.cs
// SPDX-License-Identifier: MIT

using System.Runtime.CompilerServices;
using System.Security.Cryptography;
using System.Text;

namespace AetherMesh.Storage;

/// <summary>
/// Durable key-value store backed by one file per entry in a configurable root directory.
/// Writes are atomic on the local file system: bytes go to a temp file inside the same
/// directory and are then renamed over the target. Keys are sanitized to a hex hash
/// (with the original key recoverable from a sidecar manifest) so arbitrary key strings
/// — including paths, slashes, and Unicode — round-trip safely on every host OS.
///
/// This is a simple reference impl, not a database: it doesn't compact, doesn't transact
/// across multiple keys, and has no encryption-at-rest. Hosts that need any of those
/// supply their own <see cref="IKeyValueStore"/> implementation.
/// </summary>
public sealed class FileSystemKeyValueStore : IKeyValueStore
{
    private const string EntrySuffix = ".kv";
    private const string TempSuffix = ".tmp";
    private const string KeyManifestSuffix = ".key";

    private readonly string _root;

    /// <summary>
    /// Create a store rooted at <paramref name="rootDirectory"/>. The directory is created
    /// if it does not exist. Multiple stores can share a root with disjoint <paramref name="namespace"/> values.
    /// </summary>
    public FileSystemKeyValueStore(string rootDirectory, string? @namespace = null)
    {
        ArgumentException.ThrowIfNullOrEmpty(rootDirectory);
        _root = string.IsNullOrEmpty(@namespace) ? rootDirectory : Path.Combine(rootDirectory, @namespace);
        Directory.CreateDirectory(_root);
    }

    public async Task<byte[]?> GetAsync(string key, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(key);
        var path = EntryPath(key);
        if (!File.Exists(path)) return null;
        return await File.ReadAllBytesAsync(path, cancellationToken).ConfigureAwait(false);
    }

    public async Task PutAsync(string key, byte[] value, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(key);
        ArgumentNullException.ThrowIfNull(value);

        var entry = EntryPath(key);
        var temp = entry + TempSuffix;
        await File.WriteAllBytesAsync(temp, value, cancellationToken).ConfigureAwait(false);
        File.Move(temp, entry, overwrite: true);

        var keyManifest = entry + KeyManifestSuffix;
        if (!File.Exists(keyManifest))
            await File.WriteAllTextAsync(keyManifest, key, Encoding.UTF8, cancellationToken).ConfigureAwait(false);
    }

    public Task<bool> RemoveAsync(string key, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(key);
        var entry = EntryPath(key);
        var existed = File.Exists(entry);
        if (existed)
        {
            File.Delete(entry);
            var manifest = entry + KeyManifestSuffix;
            if (File.Exists(manifest)) File.Delete(manifest);
        }
        return Task.FromResult(existed);
    }

    public Task<bool> ContainsAsync(string key, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(key);
        return Task.FromResult(File.Exists(EntryPath(key)));
    }

    public async IAsyncEnumerable<string> ListKeysAsync([EnumeratorCancellation] CancellationToken cancellationToken = default)
    {
        if (!Directory.Exists(_root)) yield break;

        foreach (var manifest in Directory.EnumerateFiles(_root, "*" + EntrySuffix + KeyManifestSuffix))
        {
            cancellationToken.ThrowIfCancellationRequested();
            string original;
            try
            {
                original = await File.ReadAllTextAsync(manifest, Encoding.UTF8, cancellationToken).ConfigureAwait(false);
            }
            catch (IOException)
            {
                continue;
            }
            yield return original;
        }
    }

    private string EntryPath(string key)
        => Path.Combine(_root, HashKey(key) + EntrySuffix);

    private static string HashKey(string key)
    {
        // SHA-256 → hex makes a filesystem-safe, fixed-length filename for any input.
        Span<byte> hash = stackalloc byte[32];
        if (!SHA256.TryHashData(Encoding.UTF8.GetBytes(key), hash, out _))
            throw new InvalidOperationException("SHA-256 hash computation failed");
        return Convert.ToHexString(hash).ToLowerInvariant();
    }
}

File: typescript/src/storage/FileSystemKeyValueStore.ts
/**
 * Durable {@link KeyValueStore} backed by one file per entry in a
 * configurable root directory. Writes are atomic on the local file system:
 * bytes go to a temp file inside the same directory and are then renamed
 * over the target. Keys are sanitised to a hex SHA-256 hash (with the
 * original key recoverable from a sidecar manifest) so arbitrary key
 * strings — including paths, slashes, and Unicode — round-trip safely on
 * every host OS.
 *
 * This is a simple reference impl, not a database: it doesn't compact,
 * doesn't transact across multiple keys, and has no encryption-at-rest.
 * Hosts that need any of those wrap the inner store with
 * {@link EncryptedKeyValueStore} or supply their own implementation.
 *
 * Mirrors the C# {@code FileSystemKeyValueStore} in
 * src/AetherMesh.Storage/FileSystemKeyValueStore.cs.
 *
 * SPDX-License-Identifier: MIT
 */
import { createHash } from "node:crypto";
import { mkdir, readdir, readFile, rename, rm, stat, writeFile } from "node:fs/promises";
import { join } from "node:path";
import { KeyValueStore } from "./KeyValueStore.js";

const ENTRY_SUFFIX = ".kv";
const TEMP_SUFFIX = ".tmp";
const KEY_MANIFEST_SUFFIX = ".key";

export class FileSystemKeyValueStore implements KeyValueStore {
  private readonly root: string;
  /** Lazily-resolved promise that the root directory exists. */
  private readyPromise: Promise<void> | null = null;

  /**
   * Create a store rooted at {@code rootDirectory}. The directory is
   * created on the first read/write if it does not exist. Multiple stores
   * can share a root with disjoint {@code namespace} values.
   */
  constructor(rootDirectory: string, namespace?: string) {
    if (!rootDirectory) throw new Error("rootDirectory cannot be empty");
    this.root = namespace ? join(rootDirectory, namespace) : rootDirectory;
  }

  private async ensureRoot(): Promise<void> {
    if (this.readyPromise === null) {
      this.readyPromise = mkdir(this.root, { recursive: true }).then(() => undefined);
    }
    return this.readyPromise;
  }

  async get(key: string): Promise<Uint8Array | null> {
    if (!key) throw new Error("key cannot be empty");
    await this.ensureRoot();
    const path = this.entryPath(key);
    try {
      const buf = await readFile(path);
      return new Uint8Array(buf);
    } catch (err: any) {
      if (err && err.code === "ENOENT") return null;
      throw err;
    }
  }

  async put(key: string, value: Uint8Array): Promise<void> {
    if (!key) throw new Error("key cannot be empty");
    if (value === null || value === undefined) {
      throw new Error("value cannot be null/undefined");
    }
    await this.ensureRoot();

    const entry = this.entryPath(key);
    const temp = entry + TEMP_SUFFIX;
    await writeFile(temp, value);
    // rename is atomic on the same filesystem (POSIX) and overwrite-safe
    // on Windows since Node 14.
    await rename(temp, entry);

    const keyManifest = entry + KEY_MANIFEST_SUFFIX;
    try {
      await stat(keyManifest);
    } catch (err: any) {
      if (err && err.code === "ENOENT") {
        await writeFile(keyManifest, key, { encoding: "utf8" });
      } else {
        throw err;
      }
    }
  }

  async remove(key: string): Promise<boolean> {
    if (!key) throw new Error("key cannot be empty");
    await this.ensureRoot();
    const entry = this.entryPath(key);
    let existed = false;
    try {
      await stat(entry);
      existed = true;
    } catch (err: any) {
      if (!(err && err.code === "ENOENT")) throw err;
    }
    if (existed) {
      await rm(entry, { force: true });
      const manifest = entry + KEY_MANIFEST_SUFFIX;
      await rm(manifest, { force: true });
    }
    return existed;
  }

  async contains(key: string): Promise<boolean> {
    if (!key) throw new Error("key cannot be empty");
    await this.ensureRoot();
    try {
      await stat(this.entryPath(key));
      return true;
    } catch (err: any) {
      if (err && err.code === "ENOENT") return false;
      throw err;
    }
  }

  async listKeys(prefix?: string): Promise<string[]> {
    await this.ensureRoot();
    let entries: string[];
    try {
      entries = await readdir(this.root);
    } catch (err: any) {
      if (err && err.code === "ENOENT") return [];
      throw err;
    }
    const out: string[] = [];
    for (const name of entries) {
      if (!name.endsWith(ENTRY_SUFFIX + KEY_MANIFEST_SUFFIX)) continue;
      try {
        const original = await readFile(join(this.root, name), "utf8");
        if (prefix && !original.startsWith(prefix)) continue;
        out.push(original);
      } catch {
        // tolerate transient read errors on enumeration
        continue;
      }
    }
    return out;
  }

  private entryPath(key: string): string {
    return join(this.root, hashKey(key) + ENTRY_SUFFIX);
  }
}

function hashKey(key: string): string {
  return createHash("sha256").update(key, "utf8").digest("hex");
}
